### PR TITLE
Changes to discover and perform azdata update

### PR DESCRIPTION
**Pull request by ranasaria**
This PR does the following:
1. Adds code to discover if an upgrade is available to the installed version of azdata.
2. Performs the upgrade if the user requests it.
3. Adds tests for the new functionality.

Testing:
1. All extension unit tests.
2. End to end manual testing, to verify successful, discovery and upgrade from old installed version to latest version.

### CODE REVIEW

File: extensions/azdata/src/test/common/httpClient.test.ts [hunk 1]
```diff
@@ -20,20 +20,26 @@ describe('HttpClient', function (): void {
 
 	before(function (): void {
 		outputChannelMock = TypeMoq.Mock.ofType<vscode.OutputChannel>();
+		if (process.env.SendOutputChannelToConsole) {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/test/common/httpClient.test.ts]: Done

File: extensions/azdata/src/test/azdata.test.ts [hunk 2]
```diff
@@ -3,22 +3,26 @@
...
-import * as sinon from 'sinon';
 import * as childProcess from '../common/childProcess';
-import * as should from 'should';
 import * as utils from '../common/utils';
-import * as nock from 'nock';
+import * as loc from '../localizedConstants';
 
+const outputChannelMock = TypeMoq.Mock.ofType<vscode.OutputChannel>();
+const oldAzdata = new azdata.AzdataTool('', new SemVer('0.0.0'), outputChannelMock.object);
+if (process.env.SendOutputChannelToConsole) {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts]: Done

File: extensions/azdata/src/test/azdata.test.ts [hunk 3]
```diff
@@ -47,26 +51,247 @@ describe('azdata', function () {
...
 	});
 
-	// TODO: Install not implemented on linux yet
-	describe('downloadAndInstallAzdata', function (): void {
-		it('successful download & install', async function (): Promise<void> {
-			sinon.stub(childProcess, 'executeCommand').returns(Promise.resolve({ stdout: '', stderr: '' }));
-			if (process.platform === 'linux') {
-				sinon.stub(childProcess, 'executeSudoCommand').returns(Promise.resolve({ stdout: '', stderr: '' }));
+	describe('installAzdata', function (): void {
+		it('successful install - win32', async function (): Promise<void> {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts]: Done

File: extensions/azdata/src/test/azdata.test.ts [hunk 4]
```diff
@@ -47,26 +51,247 @@ describe('azdata', function () {
...
+			} finally {
+				Object.defineProperty(process, 'platform', { value: originalPlatform });
+			}
+		});
+
+		it.skip('successful upgrade - linux', async function (): Promise<void> {
+			const originalPlatform = process.platform;
+			try {
+				Object.defineProperty(process, 'platform', { value: 'darwin' });
+				// const executeCommandStub = sinon.stub(childProcess, 'executeCommand').callsFake(async (command: string, _args: string[]) => {
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts]: done

File: extensions/azdata/src/test/azdata.test.ts [hunk 5]
```diff
@@ -47,26 +48,251 @@ describe('azdata', function () {
...
 
-	// TODO: Install not implemented on linux yet
-	describe('downloadAndInstallAzdata', function (): void {
-		it('successful download & install', async function (): Promise<void> {
-			sinon.stub(childProcess, 'executeCommand').returns(Promise.resolve({ stdout: '', stderr: '' }));
-			if (process.platform === 'linux') {
-				sinon.stub(childProcess, 'executeSudoCommand').returns(Promise.resolve({ stdout: '', stderr: '' }));
+	describe('installAzdata', function (): void {
+		it('successful install', async function (): Promise<void> {
+			switch (process.platform) {
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/test/azdata.test.ts:53]: I'd suggest splitting these each up into their own test that only runs on the appropriate platform. Lot easier to read.

File: extensions/azdata/src/test/common/httpClient.test.ts [hunk 6]
```diff
@@ -81,34 +80,30 @@ describe('HttpClient', function (): void {
...
 			nock('https://127.0.0.1')
-				.get(`/${filename}`)
-				.reply(200);
-			const receivedFilename = await HttpClient.getFilename(`https://127.0.0.1/${filename}`, outputChannelMock.object);
-
-			should(receivedFilename).equal(filename);
+				.get('/arbitraryFile')
+				.replyWithFile(200, __filename);
+			const receivedContents = await HttpClient.getTextContent(`https://127.0.0.1/arbitraryFile`, outputChannelMock.object);
+			console.log(`received fileContents:${receivedContents}`);
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/test/common/httpClient.test.ts:89]: remove

File: extensions/azdata/src/azdata.ts [hunk 7]
```diff
@@ -49,10 +55,19 @@ class AzdataTool implements IAzdataTool {
 	}
 }
 
+export type AzdataLatestVersionInfo = {
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/azdata.ts:58]: This is specific to brew right? Please indicate that in the name somehow - something like `AzdataDarwinPackageVersionInfo`

File: extensions/azdata/src/localizedConstants.ts [hunk 8]
```diff
@@ -12,19 +12,28 @@ export function downloadingProgressMb(currentMb: string, totalMb: string): strin
...
+export const no = localize('azdata.no', "No");
 export function downloadingTo(name: string, location: string): string { return localize('azdata.downloadingTo', "Downloading {0} to {1}", name, location); }
 export function executingCommand(command: string, args: string[]): string { return localize('azdata.executingCommand', "Executing command \"{0} {1}\"", command, args?.join(' ')); }
-export function stdoutOutput(stdout: string): string { return localize('azdat.stdoutOutput', "stdout : {0}", stdout); }
-export function stderrOutput(stderr: string): string { return localize('azdat.stderrOutput', "stderr : {0}", stderr); }
+export function stdoutOutput(stdout: string): string { return localize('azdata.stdoutOutput', "stdout : {0}", stdout); }
+export function stderrOutput(stderr: string): string { return localize('azdata.stderrOutput', "stderr : {0}", stderr); }
+export const checkingLatestAzdataVersion = localize('azdata.checkingLatestAzdataVersion', "Checking for latest version of azdata");
 export function gettingFilenameOfUrl(url: string): string { return localize('azdata.gettingFilenameOfUrl', "Getting filename of resource at URL {0}", url); }
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/localizedConstants.ts]: done

File: extensions/azdata/src/test/azdata.test.ts [hunk 9]
```diff
@@ -52,67 +49,20 @@ describe('azdata', function () {
...
 					break;
 			}
 		});
 
 		if (process.platform === 'win32') {
 			it('unsuccessful download - win32', async function (): Promise<void> {
-				nock(azdata.azdataHostname)
-					.get(`/${azdata.azdataUri}`)
-					.reply(404);
+				sinon.stub(HttpClient, 'downloadFile').rejects();
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/test/azdata.test.ts:65]: Why did you change this? I would prefer we keep the test handling the actual HTTP errors.

File: extensions/azdata/src/test/azdata.test.ts [hunk 10]
```diff
@@ -296,3 +144,165 @@ describe('azdata', function () {
...
+	const executeCommandStub = sinon.stub(childProcess, 'executeCommand').callsFake(async (command: string, _args: string[]) => {
+		should(command).be.equal('add-apt-repository');
+		return { stdout: 'success', stderr: '' };
+	});
+	let callNumber = 0;
+	const executeSudoCommandStub = sinon.stub(childProcess, 'executeSudoCommand').callsFake(async (command: string) => {
+		callNumber++;
+		switch (callNumber) {
+			case 3:
+				should(command).match('/^curl/');
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/test/azdata.test.ts:180]: I think this level of testing is too much and not providing any value. Just verifying that it was called at least once is enough - since we're not actually calling the methods on the platform then we should just rely on the implementer knowing what commands need to be executed to handle this correctly.

And FYI - if this was something you needed to actually do you can just use the onCall functions to do this instead of keeping counters yourself.

`stub.onCall(3).calledWith(...)`

File: extensions/azdata/src/azdata.ts [hunk 11]
```diff
@@ -153,16 +223,85 @@ async function findAzdataWin32(outputChannel: vscode.OutputChannel): Promise<IAz
...
+ * @param outputChannel Channel used to display diagnostic information
+ */
+async function discoverLatestAzdataVersionFromJson(outputChannel: vscode.OutputChannel): Promise<SemVer> {
+	// get version information for current platform from http://aka.ms/azdata/release.json
+	const fileContents = await HttpClient.getTextContent(`${azdataHostname}/${azdataReleaseJson}`, outputChannel);
+	let azdataReleaseInfo;
+	try {
+		azdataReleaseInfo = JSON.parse(fileContents);
+	} catch (e) {
+		throw Error(`failed to parse the JSON of contents at: ${azdataHostname}/${azdataReleaseJson}, error:${getErrorMessage(e)}`);
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/azdata.ts:265]: Include the downloaded JSON in the message too

File: extensions/azdata/src/azdata.ts [hunk 12]
```diff
@@ -153,16 +223,85 @@ async function findAzdataWin32(outputChannel: vscode.OutputChannel): Promise<IAz
...
+ */
+async function discoverLatestStableAzdataVersionDarwin(outputChannel: vscode.OutputChannel): Promise<SemVer> {
+	// set brew tap to azdata-cli repository
+	await executeCommand('brew', ['tap', 'microsoft/azdata-cli-release'], outputChannel);
+	let brewInfoAzdataCliJson;
+	const brewInfoOutput = (await executeCommand('brew', ['info', 'azdata-cli', '--json'], outputChannel)).stdout;
+	try {
+		brewInfoAzdataCliJson = JSON.parse(brewInfoOutput);
+	} catch (e) {
+		throw Error(`failed to parse the JSON contents output of: 'brew info azdata-cli --json', error:${getErrorMessage(e)}`);
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/azdata.ts:284]: Include the downloaded JSON in the message too

File: extensions/azdata/src/azdata.ts [hunk 13]
```diff
@@ -153,16 +223,86 @@ async function findAzdataWin32(outputChannel: vscode.OutputChannel): Promise<IAz
...
+/**
+ * Gets the latest azdata version for MacOs clients
+ * @param outputChannel Channel used to display diagnostic information
+ */
+async function discoverLatestStableAzdataVersionDarwin(outputChannel: vscode.OutputChannel): Promise<SemVer> {
+	// set brew tap to azdata-cli repository
+	await executeCommand('brew', ['tap', 'microsoft/azdata-cli-release'], outputChannel);
+	let brewInfoAzdataCliJson;
+	const brewInfoOutput = (await executeCommand('brew', ['info', 'azdata-cli', '--json'], outputChannel)).stdout;
+	console.log(`##############brewInfoOutput:`, brewInfoOutput);
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/azdata.ts]: done

File: extensions/azdata/src/test/azdata.test.ts (hunk 9 above)
Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts:65]: for isolation, HttpClient.test.ts already tests HttpClient APIs work correctly so just mocking at HttpClient level gives azdata tests isolation so that if HttpClient's implementation changes without corresponding api change then azdata tests are not affected.

Comment (Charles-Gagnon) [reply extensions/azdata/src/test/azdata.test.ts:65]: I kind of would have rather kept the logic and then just had a helper function that set up the mocks that both called. We don't need to completely isolate everything from everything else - especially when the code is part of the same module.

Your reasoning here isn't bad - I just think this change was unnecessary and isn't really helping anything except make more assumptions about how downloadFile behaves.

**Comment by ranasaria** [comment]
> Still have some open comments to address. Also please merge in from main - there's a couple major changes (like the logger) that you're missing here.

Yes, I have a merge from the main stashed. No way to merge back without that. However, my other review is also dependent on this one so I wanted to do all those fixes first before the merge from main is flowed all the way in. I will push the merge in later today after I complete some more testing.

File: extensions/azdata/src/test/azdata.test.ts (hunk 10 above)
Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts:180]: ok, I will keep only the most important verifications instead of verification of each call. I will user onCall. Thanks for the suggestion.

Comment (ranasaria) [reply extensions/azdata/src/test/azdata.test.ts]: done.

File: extensions/azdata/src/extension.ts [hunk 14]
```diff
@@ -39,7 +40,11 @@ export async function activate(): Promise<azdata.IExtension> {
 
 async function checkForAzdata(outputChannel: vscode.OutputChannel): Promise<IAzdataTool | undefined> {
 	try {
-		return await findAzdata(outputChannel);
+		const azdata = await findAzdata(outputChannel); // find currently installed Azdata
+		vscode.window.showInformationMessage(loc.foundExistingAzdata(azdata.path, azdata.version.raw));
+		// Don't block on this since we want the extension to finish activating without needing user input
+		checkAndUpdateAzdata(azdata, outputChannel).catch(err => vscode.window.showWarningMessage(loc.updateError(err))); //update if available and user wants it.
+		return findAzdata(outputChannel); // now again find and return the currently installed azdata
```
[the remark this answers is not in the document]
Comment (ranasaria) [reply extensions/azdata/src/extension.ts:71]: done

**Comment by ranasaria** [comment]
> > Still have some open comments to address. Also please merge in from main - there's a couple major changes (like the logger) that you're missing here.
> 
> Yes, I have a merge from the main stashed. No way to merge back without that. However, my other review is also dependent on this one so I wanted to do all those fixes first before the merge from main is flowed all the way in. I will push the merge in later today after I complete some more testing.

Merge from main done.

File: extensions/azdata/src/azdata.ts [hunk 15]
```diff
@@ -90,10 +102,19 @@ class AzdataTool implements IAzdataTool {
...
+	/**
+	 * Gets the output of running '--version' command on the azdata tool.
+	 * It also updates the cachedVersion property based on the return value from the tool.
+	 */
 	public async version(): Promise<azdataExt.AzdataOutput<string>> {
-		const output = await this.executeCommand<string>(['--version']);
-		this.toolVersion = parseVersion(output.stdout[0]);
-		return output;
+		const output = await executeCommand(`"${this.path}"`, ['--version']);
+		this.cachedVersion = new SemVer(parseVersion(output.stdout[0]));
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/azdata.ts:111]: output.stdout is a single string so indexing into it doesn't seem correct.

File: extensions/azdata/src/common/httpClient.ts [hunk 16]
```diff
@@ -17,13 +17,31 @@ export namespace HttpClient {
...
+	export async function getTextContent(url: string): Promise<string> {
+		Logger.log(loc.gettingTextContentsOfUrl(url));
+		return await download(url);
+	}
+
+	/**
 	 * @returns Full path to the downloaded file
 	 */
-	export function download(downloadUrl: string, targetFolder: string): Promise<string> {
+	function download(downloadUrl: string, targetFolder?: string): Promise<string> {
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/common/httpClient.ts:39]: Update doc to indicate that if targetFolder isn't defined it'll return the contents

File: extensions/azdata/src/extension.ts [hunk 17]
```diff
@@ -85,7 +92,8 @@ function throwIfNoAzdata(): void {
 
 async function checkForAzdata(): Promise<IAzdataTool | undefined> {
 	try {
-		return await findAzdata();
+		const azdata = await findAzdata(); // find currently installed Azdata
+		vscode.window.showInformationMessage(loc.foundExistingAzdata(azdata.path, azdata.cachedVersion.raw));
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/extension.ts:96]: Just log this to output channel

File: extensions/azdata/src/extension.ts [hunk 18]
```diff
@@ -85,7 +92,8 @@ function throwIfNoAzdata(): void {
 
 async function checkForAzdata(): Promise<IAzdataTool | undefined> {
 	try {
-		return await findAzdata();
+		const azdata = await findAzdata(); // find currently installed Azdata
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/extension.ts:95]: You're never returning this

File: extensions/azdata/src/localizedConstants.ts [hunk 19]
```diff
@@ -12,19 +12,28 @@ export function downloadingProgressMb(currentMb: string, totalMb: string): strin
 export const downloadFinished = localize('azdata.downloadFinished', "Download finished");
 export const install = localize('azdata.install', "Install");
 export const installingAzdata = localize('azdata.installingAzdata', "Installing azdata...");
+export const upgradingAzdata = localize('azdata.upgradingAzdata', "Upgrading azdata...");
 export const azdataInstalled = localize('azdata.azdataInstalled', "azdata was successfully installed. Restarting Azure Data Studio is required to complete configuration - features will not be activated until this is done.");
+export const azdataUpgraded = localize('azdata.azdataUpgraded', "azdata was successfully upgraded.");
```
[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply extensions/azdata/src/localizedConstants.ts:16]: Ping - please add the version here

File: extensions/azdata/src/localizedConstants.ts [hunk 20]
```diff
@@ -7,25 +7,30 @@ import * as nls from 'vscode-nls';
...
+export const stderrOutput = (stderr: string): string => localize('azdata.stderrOutput', "stderr : {0}", stderr);
+export const checkingLatestAzdataVersion = localize('azdata.checkingLatestAzdataVersion', "Checking for latest version of azdata");
+export const gettingTextContentsOfUrl = (url: string): string => localize('azdata.gettingTextContentsOfUrl', "Getting text contents of resource at URL {0}", url);
+export const foundAzdataVersionToUpgradeTo = (version: string): string => localize('azdata.versionForUpgrade', "Found version {0} that azdata-cli can be upgraded to.", version);
+export const promptForAzdataUpgrade = (version: string): string => localize('azdata.promptForAzdataUpgrade', "An updated version of azdata ( {0} ) is available, do you wish to install it now?", version);
+export const couldNotFindAzdata = (err: any): string => localize('azdata.couldNotFindAzdata', "Could not find azdata. Error : {0}", err.message ?? err);
+export const couldNotFindAzdataWithPrompt = localize('azdata.couldNotFindAzdataWithPrompt', "Could not find azdata, install it now? If not then some features will not be able to const.");
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/localizedConstants.ts:29]: Something got messed up with the string here - `If not then some features will not be able to function.`

File: extensions/azdata/src/test/azdata.test.ts [hunk 21]
```diff
@@ -40,26 +43,227 @@ describe('azdata', function () {
...
+				case 'linux':
+					await testLinuxUnsuccessfulInstall();
+					break;
+			}
+		});
+	});
+
+	describe('upgradeAzdata', function (): void {
+		beforeEach(function (): void {
+			// const mock = TypeMoq.Mock.ofInstance(azdata.discoverLatestAvailableAzdataVersion)
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/test/azdata.test.ts:86]: remove

File: extensions/azdata/src/typings/azdata-ext.d.ts [hunk 22]
```diff
@@ -147,6 +147,20 @@ declare module 'azdata-ext' {
 		}
 	}
 
+	export interface PostgresServerShowResult {
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/typings/azdata-ext.d.ts:150]: Remove - appears to be duplicate of below

Comment (ranasaria) [reply extensions/azdata/src/typings/azdata-ext.d.ts:150]: yeah - looks like an auto-merge issue

File: extensions/azdata/src/extension.ts (hunk 17 above)
Comment (ranasaria) [reply extensions/azdata/src/extension.ts:96]: I think this was addressed in other PR but will do it here also.

File: extensions/azdata/src/localizedConstants.ts (hunk 20 above)
Comment (ranasaria) [reply extensions/azdata/src/localizedConstants.ts:29]: can you elaborate?

**Comment by ranasaria** [comment]
With regards to https://github.com/microsoft/azuredatastudio/pull/11906#discussion_r478537295, the install and upgrade sequences are followed by findAzdata which prints out the version found. Without finding Azdata installed we do not know the version to include in these messages.

File: extensions/azdata/src/localizedConstants.ts (hunk 20 above)
Comment (Charles-Gagnon) [reply extensions/azdata/src/localizedConstants.ts:29]: You changed the text to ` If not then some features will not be able to const.` which seems to be a typo. That should be reverted.

Comment (ranasaria) [reply extensions/azdata/src/localizedConstants.ts:29]: Need more details on, in any case couldNotFindAzdataWithPrompt is in commented out code which is addressed in second review so let's look at it there.

File: extensions/azdata/src/azdata.ts (hunk 15 above)
Comment (ranasaria) [reply extensions/azdata/src/azdata.ts]: done

File: extensions/azdata/src/common/httpClient.ts (hunk 16 above)
Comment (ranasaria) [reply extensions/azdata/src/common/httpClient.ts:40]: done

File: extensions/azdata/src/extension.ts (hunk 17 above)
Comment (ranasaria) [reply extensions/azdata/src/extension.ts]: removed as findAzdata already logs to output channel

File: extensions/azdata/src/extension.ts (hunk 18 above)
Comment (ranasaria) [reply extensions/azdata/src/extension.ts]: Good catch. Fixed.

File: extensions/azdata/src/localizedConstants.ts (hunk 20 above)
Comment (ranasaria) [reply extensions/azdata/src/localizedConstants.ts:29]: Ah ok.

Comment (ranasaria) [reply extensions/azdata/src/localizedConstants.ts:29]: Done. Good catch.

File: extensions/azdata/src/common/httpClient.ts [hunk 23]
```diff
@@ -17,13 +17,32 @@ export namespace HttpClient {
...
+	 * @returns a promise to a string that has the contents of document at the provided url
+	 */
+	export async function getTextContent(url: string): Promise<string> {
+		Logger.log(loc.gettingTextContentsOfUrl(url));
+		return await download(url);
+	}
+
+	/**
+	 * @returns Full path to the downloaded file.
+	 * If the targetFolder is not defined then it returns the contents of the downloaded file.
```
Comment (Charles-Gagnon) [inline extensions/azdata/src/common/httpClient.ts:38]: Please use the @param annotations for the parameters - see other functions for examples.
